AVIF images do not render in iOS Simulator

nmoucht opened the pull request:
#### ca9f53c112bce3ac9330e219a7d35a00b86e8fc0
<pre>
AVIF images do not render in iOS Simulator
<a href="https://bugs.webkit.org/show_bug.cgi?id=247831">https://bugs.webkit.org/show_bug.cgi?id=247831</a>
rdar://102534920

Reviewed by Said Abou-Hallawa.

In iOS simulator we are not decoding AVIF images because in
defaultSupportedImageTypes we are filtering the image types
based on CGImageSourceCopyTypeIdentifiers. Currently this
function is not returning public.avif so we do not decode the
image. To workaround this, we add the avif related types after
filtering the image types. I filed rdar://104940377 to get
CGImageSourceCopyTypeIdentifiers returning public.avif in
the iOS simulator.

* Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
(WebCore::defaultSupportedImageTypes):

Canonical link: <a href="https://commits.webkit.org/259819@main">https://commits.webkit.org/259819@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c455d757fbf67c943f964da06bb2889d6132d94

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105923 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115116 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175234 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6180 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98150 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114868 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12491 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95464 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40036 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94356 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27108 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81701 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/95603 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8253 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28461 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/94906 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6058 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8736 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5042 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30465 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14366 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48002 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/103654 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10289 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25697 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3650 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
```diff
@@ -57,10 +57,6 @@ const MemoryCompactLookupOnlyRobinHoodHashSet<String>& defaultSupportedImageType
-#endif
-#if HAVE(AVIF) || USE(AVIF)
-            "public.avif"_s,
-            "public.avis"_s,
```
I think we should keep these entries in this list. This is the list of the default image types WebKit supports for CG ports. I think adding USE(AVIF) was wrong  and should be removed.

nmoucht replied:
Fixed.

shallawa commented on Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
```diff
@@ -78,6 +74,9 @@ const MemoryCompactLookupOnlyRobinHoodHashSet<String>& defaultSupportedImageType
         }
+        // rdar://104940377 Workaround for CGImageSourceCopyTypeIdentifiers not returning AVIF for iOS simulator
+        filtered.add("public.avif"_s);
+        filtered.add("public.avis"_s);
```
This code will be enabled on down level macOS (i.e. Monetary) which is incorrect. I think it should be enabled only on iOS simulator 16.0.

nmoucht replied:
Fixed.

shallawa commented on Source/WTF/wtf/PlatformUse.h:
```diff
@@ -376,6 +376,10 @@
 
+#if PLATFORM(IOS_SIMULATOR) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 160000
+#define USE_AVIF_FILE_TYPES_FOR_CG_BUG 1
+#endif
```
Should not this macro be in PlatfromHave.h:

```
#define HAVE_CG_IMAGE_SOURCE_AVIF_IMAGE_TYPES_BUG 1
```

or something similar?

nmoucht replied:
Fixed.

nmoucht committed:
AVIF images do not render in iOS Simulator
https://bugs.webkit.org/show_bug.cgi?id=247831
rdar://102534920

Reviewed by Said Abou-Hallawa.

In iOS simulator we are not decoding AVIF images because in
defaultSupportedImageTypes we are filtering the image types
based on CGImageSourceCopyTypeIdentifiers. Currently this
function is not returning public.avif so we do not decode the
image. To workaround this, we add the avif related types after
filtering the image types. I filed rdar://104940377 to get
CGImageSourceCopyTypeIdentifiers returning public.avif in
the iOS simulator.

* Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
(WebCore::defaultSupportedImageTypes):

Canonical link: https://commits.webkit.org/259819@main
